### ISSUE TITLE
Add naive core interpreter so TH and GHCi work even with ABI changes

### ISSUE THREAD

**Issue by Ericson2314** [ghc-proposals/ghc-proposals#162]
external interpreters and `--target` help with the cross case, but that still leaves the ABI compatibility problem. The end result for GHC development itself is the same: TH and GHCi are not availible in certain bootstrapping scenarios, with restricts the way GHC and core libs must be developed in myriad obscure and highly new-contributor-hostile ways.

Also, it seems like a dedicated core interpeter would be a good thing to have in general as a "living quasi-specification", including giving something authoritative for people to read, oracle for tests, etc.

**Comment by goldfirere** [comment]
Apologies if I'm being dense, but I really do not understand what this is proposing. What is `--target`? What is the "ABI compatibility problem"? How do `--target` and external intepreters help? What don't they do? How would a Core interpreter help? What's "highly hostile" about today's setup?

**Comment by angerman** [comment]
@goldfirere, you need to look at this through the cross compilation lenses.

- `-target` refers to ghc being eventually able to produce code for multiple targets from the same compiler by providing the relevant `-target` flag to ghc. Similar to what `clang` allows.

- `-fexternal-interpreter` runs a secondary process (called `iserv`) as the interpreter. This allows a stage1 GHC to evaluate TH splices via the external interpreter process.

TH in its current (unrestricted) form is a major hurdle to get cross compilation working with GHC.

**Comment by Ericson2314** [comment]
@angerman tells me about the source plugins new to GHC 8.6. These are very cool, and I think better embody the current implementation strategy. A plugin by nature accesses far too vast an API to be worth the manual virtualization and mashalling the interpreter method uses. Using the old GHC (worst case) to make code ABI-compatible with the new GHC is the necessary price to access all the bells and whistles of the compiler library.

**Comment by simonmar** [comment]
Fun fact: GHCi was originally implemented as a Core interpreter. But that implementation never made it as far as a release, because we couldn't get it to fully work. The sticking point was unboxed types (and would probably be unboxed tuples these days, too). It would be possible if the interpreter didn't have to interact directly with compiled code, but for GHCi it does.  Without that constraint the interpreter could choose its own representation for unboxed values (i.e. it could use boxed ones). 

But I presume the idea here is to interpret everything, including the base package. For that you would need some way to persist the Core when compiling modules, and load it into the interpreter.

**Comment by Ericson2314** [comment]
Yes persisting the core would be the eventual goal. How different is core from the "unfoldings" we already persist for downstream inlining?

But as a stop gap, we can implemenent multi-package source loading. That is super wanted for GHCi and more incremental `ghc --make` anyways.

**Comment by goldfirere** [comment]
It seems this conversation is useful to the cognoscenti in this area. However, unless I'm in the minority here, I'll need quite a bit more text (i.e. a full proposal) to understand this when it comes to the committee. I appreciate the effort in https://github.com/ghc-proposals/ghc-proposals/issues/162#issuecomment-414562378, but I still feel the need for concrete examples showing where, exactly, the current story founders.

If this preliminary conversation is useful, no need to spend lots of time writing it up now, but (as a committee member) I'd very much like to see this idea in the context of a full proposal.

**Comment by treeowl** [comment]
@simonmar, could you explain what makes unboxed tuples so very hard for an interpreter?

**Comment by simonmar** [comment]
@treeowl I'm not sure I can recall exactly where we ran into trouble. But consider that you need to `unsafeCoerce' things all over the place because the interpreter (in Haskell) needs to invoke native compiled functions and pass arguments and collect results in the correct way. If you don't need to invoke native compiled functions from the interpreter then it's much easier, because you can choose representations.

**Comment by Ericson2314** [comment]
And yes, it's precisely in avoiding calling "native compiled functions" that we escape the ABI restrictions that complicate bootstrapping.

**Comment by simonmar** [comment]
Right, so I think the problem is that you need an infinite number of different return frames because there are an infinite number of return conventions for functions returning unboxed tuples. You could always generate them at runtime - rather like the way we generate info tables for constructors, which also have a little bit of code - but the return frames would be significantly more complex, and you have to do it for each platform we support.

**Comment by Ericson2314** [comment]
There is a similar problem calling foreign code, but at least that doesn't evolve in lockstep to the GHC ABIs. And hopefully libffi would make that easy enough.

**Comment by mchakravarty** [comment]
Isn't the simplest way to think about this to say, we add a new target called `core` (which is always available). It never calls native code (or rather Core is the native code for that target) and you cannot link it — hence, it is only useful for being loaded for GHCi and TH.

**Comment by Ericson2314** [comment]
Yes, core as a "way" does strike me as better for those reasons. Regardless, the actual work here is not producing core---we already can do that!---but evaluating it so that nothing beyond core is needed.

**Comment by chrisdone** [comment]
You're in for an adventure.

I made a start on a "pure-Haskell" STG interpreter a year ago: https://github.com/chrisdone/prana It could run [some simple functions](https://github.com/chrisdone/prana/tree/master/prana-interpreter/test/assets).

As Simon suggested, it interpreted everything all the way down: base, ghc-prim's primops, etc. Otherwise you have to "FFI" from your interpreter to native GHC's runtime representation.

If you're looking for a "feasibility study" of how hard this would be to do, you can consider that project a modern (ish) analysis. It also doesn't do any fancy optimizations, every step is the dumbest thing possible. I even put tasks in a table with my appraisal of how difficult each was or was expected to be. Exceptions, for example, raise an interesting design challenge. Another open challenge is FFI support. 

I would recommend interpreting **STG, not Core**. Core is full of assumptions and missing things. STG is the actual "ready to interpret" type. I tried Core first and hit a dead end. STG is also simpler, as it has restrictions that all thunks are explicit, there are only lets, no lambdas (ostensibly, but I seen 'em), etc. Quoting myself from 2019:

> I think this mailing list thread turns out to be an x/y problem. I need STG. 
>
> I ended up doing lots of cleaning steps to generate a more well-formed 
Core. I learned that the Core that comes from the desugarer is 
incomplete and realized today that I was duplicating work done by the 
transformation that converts Core to STG. I'd initially avoided STG, 
thinking that Core would suffice, but it seems that STG is the AST 
that I was converging on anyway. It seems that STG will have all class 
methods generated, wrappers generated (and inlined as appropriate), 
primops and ffi calls are saturated, etc. no Types or coercions, etc. 
It even has info about whether closures can be updated and how. 

Just running the "lambda calculus with thunks" is one thing, but if you want to genuinely interpret GHC Haskell it's a much broader surface area.

**Comment by int-index** [comment]
I agree that an STG interpreter would be a better fit for this than a Core interpreter. And, quite excitingly, @csabahruska has developed one (with an entirely different motivation!)

I’d love to see Csaba’s work go upstream and applied to solve the problem of making TH available in more scenarios (e.g. in stage1 GHC, so that GHC itself could use TH, that would be terrific).

Furthermore, I think this would help DTs. There’s one thing common between TH and DTs: compile-time evaluation. Currently, type-level computation is done with type families, and their performance is quite bad. Instead, what we should aim for is that promoted functions have the same space/time complexity at the type level as they do at the term-level. At least when they are applied to plain data (i.e. no type variables). And one way to achieve that would be to compile them to STG, and then use an STG interpreter. We could even do that before DTs, to speed up type families.

**Comment by Ericson2314** [comment]
Makes sense. I wish we could do an optional typed STG, and then have STG lint too.

More on topic, it's occurred to me that it's less that the existing bytecode interpreter is unfit for this purpose, than trying to do the FFI that @chrisdone mentions.

If in conjunction with https://gitlab.haskell.org/ghc/ghc/-/issues/18954 we can get enough core/stg/whatever that we're never mixing compiled and interpreted code, maybe the existing bytecode is fine. CC @joshmeredith.

(Also N.B. @luite is making the bytecode generation itself work from STG.)

**Comment by Ericson2314** [comment]
https://www.patreon.com/posts/external-stg-49857800 it looks like we might have this now! CC @csabahruska
